Update events period method to check for invalid values

adbridge opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
The period method currently allows any ms value: 
positive, negative and zero. 

A negative value means dispatch a non periodic event. ie just once. 
0 is unspecified behaviour. 

This commit forces the user to use either positive periods or a new constant, non_periodic
and should any other value be provided will default to non_periodic.

A Greentea test case is also provided to check this works as
expected.

I'm classifying this as a feature change rather than major because allowing a zero value was actually a bug.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
If the user provides a 0 value this will now behave differently. Previously it would have run for a period of 
time as if this was an indefinite event before eventually (with a timer wrap) have unspecified behaviour.
Now the event period will default instead to non periodic.

To specify a non periodic event a user could previously use any negative value. Whilst this will still work (due a default
being set on invalid values), the user is encouraged to use the new constant, 'non_periodic' to show that this is what was
intended.


#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
Use periods that are > 0ms or non_periodic constant.


### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
Should be picked up automatically in the doxygen
----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

adbridge committed:
Update events period method to check for invalid values

The period method currently allows any ms value positive, negative
and zero. A negative value means dispatch a non periodic event
ie just once. 0 is unspecified behaviour. This commit forces the
user to use either positive periods or a new constant non_periodic
and should any other value be provided will default to
non_periodic.

A Greentea test case is also provided to check this works as
expected.

0xc0170 commented on events/tests/TESTS/events/queue/main.cpp:
```diff
@@ -536,8 +612,8 @@ const Case cases[] = {
-
-};
+    Case("Testing event period values", event_period_tests)
+};      
```
remove the spaces

0xc0170 commented on events/tests/TESTS/events/queue/main.cpp:
```diff
@@ -508,6 +508,82 @@ void static_events_queue_test()
+    // should be 1
+    TEST_ASSERT_EQUAL(1, event_period.counter);
+
+    // Test an event with an invalid -ve period.
```
`-ve` what does this mean? should it be -10ms?

adbridge replied:
-ve = negative value for the period

evedon commented on events/include/events/Event.h:
```diff
@@ -140,12 +141,20 @@ class Event<void(ArgTs...)> {
      */
     void period(duration p)
     {
+        MBED_ASSERT(p > duration(0) || p == non_periodic);
```
`MBED_ASSERT` might be too harsh here, especially for negative values which are handled fine. Perhaps use a warning and say it will be enforced in a future release.

adbridge replied:
Will change to MBED_WARNING .

evedon commented on events/tests/TESTS/events/queue/main.cpp:
```diff
@@ -508,6 +508,82 @@ void static_events_queue_test()
     TEST_ASSERT_EQUAL(15, test4.counter);
 }
 
+static EventTest event_period;
```
You could declare as a local variable in `event_period_tests` instead.

adbridge replied:
This is removed in the updated version

evedon commented on events/tests/TESTS/events/queue/main.cpp:
```diff
@@ -508,6 +508,82 @@ void static_events_queue_test()
+    queue.dispatch(800); 
+    
+    // Wait 1000ms and check the event execution status
+    wait_us(1000 * 1000)    
```
I would reduce all the duration so that the test executes faster.

evedon replied:
`;` missing here too

evedon commented on events/tests/TESTS/events/queue/main.cpp:
```diff
@@ -508,6 +508,82 @@ void static_events_queue_test()
+    Event<void(int)> event1(&queue, &event_period::f0);
+    event1.delay(100ms);
+    event1.period(Event::non_periodic);
+    event1.post()
```
A `;` is missing. This should not compile

evedon commented on events/tests/TESTS/events/queue/main.cpp:
```diff
@@ -536,8 +612,8 @@ const Case cases[] = {
     Case("Testing static events queue", static_events_queue_test)
-
-};
+    Case("Testing event period values", event_period_tests)
```
A comma is missing line above

evedon commented on events/tests/TESTS/events/queue/main.cpp:
```diff
@@ -508,6 +508,82 @@ void static_events_queue_test()
+    event_period.counter = 0;
+    Event<void(int)> event2(&queue, &event_period::f0);
+    event2.delay(100ms);
+    event2.period(-10ms);
```
This should trigger the assert so I am not sure how it works. Is the assert mocked somehow?

adbridge replied:
All the compilation errors are already fixed , will submit an update soon

adbridge committed:
Updated event period handling and greentea test

non_periodic constant has been moved out of the Event class and
made static within the events namespace so that it is available
both internally within the class and externally.

The MBED_ASSERT has been changed to MBED_WARNING.

The greentea test has been updated:
1) timings reduced to make the test cases run faster
2) The call handler simplified

adbridge committed:
Fix some whitespace issues

evedon commented on events/include/events/Event.h:
```diff
@@ -67,7 +70,7 @@ class Event<void(ArgTs...)> {
             _event->id = 0;
             _event->delay = duration(0);
-            _event->period = duration(-1);
+            _event->period = events::non_periodic;
```
You don't need the namespace here

adbridge replied:
Fixed

evedon commented on events/include/events/Event.h:
```diff
@@ -140,12 +143,23 @@ class Event<void(ArgTs...)> {
+                    MBED_WARNING(MBED_ERROR_INVALID_ARGUMENT,
+                                 "Invalid period specified, defaulting to non_periodic.");
+                }
+                _event->period = events::non_periodic;                
```
You don't need the namespace here

adbridge replied:
Fixed

evedon reviewed (commented):
Two minor comments. The rest looks fine.

adbridge committed:
Remove unecessary namespace qualified

(Also minor whitespace correction)

adbridge committed:
Fix astyle issues

adbridge committed:
Fix whitespace to stop astyle complaining

adbridge committed:
Further whitespace fixes

adbridge committed:
one final whitespace change!

0xc0170 reviewed (dismissed):
events/include/events/Event.h  100644 → 100755 - please revert the file attributes change

adbridge committed:
Revert "one final whitespace change!"

This reverts commit c9a3375042a09e0dad03920806644209fa606e8b.

adbridge committed:
Minor formatting tweak to fix file permissions
